### ISSUE TITLE
Correct datetime of a deployment

### ISSUE THREAD

**Issue by damianooldoni** [inbo/camtrapdp#105]
From mail discussion about correcting specific datetimes

Qusetion of a user, on Jan 5, 2023: 

> Some time ago I brought up this issue about trying to correct date-time on one of my CTs. [...]

@MarcusRowcliffe's answer:

> I don’t believe there are functions currently available in camtraptor either for correcting times [...].
> I’ve had a quick go at creating basic functions for these tasks (with a warning that there is no error checking so very much up to you ensure sensible input). Your solution to time correction is right I think, but maybe not so easy to reuse! Here’s a suggestion for a more re-usable function:

```r
correct_time <- function(package, deploymentID, wrongTime, rightTime){
  tdiff <- as.POSIXct(rightTime, tz="UTC") - as.POSIXct(wrongTime, tz="UTC")
  i <- package$data$deployments$deploymentID == deploymentID
  package$data$deployments <- dplyr::mutate(package$data$deployments,
                                            start = start + tdiff,
                                            end = end + tdiff)
  package$data$observations <- dplyr::mutate(package$data$observations,
                                             timestamp = timestamp + tdiff)
  package$data$media <- dplyr::mutate(package$data$media,
                                      timestamp = timestamp + tdiff)
  package
}

Example usage:
```r
cpkg <- correct_time(package,
                     deploymentID = "022406d5-96a8-4192-a8d2-f86eab55df57",
                     wrongTime = "2017-01-01 00:00:00",
                     rightTime = "2018-09-01 11:04:00")
```

Thanks @MarcusRowcliffe to put me in cc of this conversation!

**Comment by peterdesmet** [comment]
As discussed with @damianooldoni and @sannegovaert today:

- [ ] name: `correct_time()`
- [ ] family: `transformation functions`
- [ ] params:
  - [ ] `x` (package)
  - [ ] `deploymentID`: allows vector of deploymentIDs
  - [ ] either `wrong_time` and `right_time` or `duration`
- [ ] functionality
  - [ ] calculate duration difference
  - [ ] shift `dep.deploymentStart` and `dep.deploymentEnd` for selected deployments
  - [ ] set `dep.timestampIssues` to `FALSE` sets 
  - [ ] shift `media.timestamp` for associated media
  - [ ] shift `obs.eventStart` and `eventEnd` for associated observations
  - [ ] recalculate `x.temporal` (see helper function)
- [ ] message (for the example timestamp, take the first deployment)
> Timestamps in selected deployments, media and observations were shifted by -02.00 hours (e.g. 2024-04-01T02:00:00 is now 2024-04-01T00:00:00).

Note 1: when reading the data, R will take into account the timezone (e.g. `+02.00`). When exporting, all times will be (should be) set to UTC.

Note 2: we said in our meeting that providing a `wrong_time` and `right_time` is the most convenient for the user, but I find it somewhat confusing that the datetimes in question are not used and don't even have to match with the data. E.g. as @damianooldoni noted:

> And an error should be returned if the new observations or media timestamp value is before the start or after the end value of the corresponding deployment.

I don't agree an error should be returned, since only the difference in time is used. I therefore prefer a simple `duration` argument, defined as a lubridate [Duration](https://lubridate.tidyverse.org/reference/Duration-class.html) and a good example to make it clearer. It will also make for easier tests.

``` r
library(lubridate)

wrong <- ymd_hms("2024-04-01T00:00:00", tz = "UTC")
right <- ymd_hms("2024-04-01T02:00:00", tz = "UTC")
as.duration(interval(wrong, right))
#> [1] "7200s (~2 hours)"

as.duration(interval(right, wrong))
#> [1] "-7200s (~-2 hours)"
```

<sup>Created on 2024-07-16 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by MarcusRowcliffe** [comment]
Thanks @damianooldoni abd @peterdesmet.

I'm happy with a duration argument (alongside a clear example to get there from reference times). I'm also happy for the function to return media / observation times outside deployment time without an error since this happens anyway in data generation (and is often why we need this function) - should be up to the user to sense check (e.g. using camtrapDensity::plot_deployment_schedule - perhaps checks could also be built into `read_camtrap_dp`?).

Given the range of problems I've seen, I'm keen to make the function sufficiently flexible to handle:
- changing only at one level of the deployment > media > hierarchy, or to cascade changes down to lower levels
- to change either deployment start, end or both

Here is a suggested outline function (names abbreviated for brevity but hopefully clear; note if `level = "med"` and `cascade = TRUE` there would need to be a further cascade to event observations via `eventID` - at the moment this would only affect media observations):
```
corrdt <- function(pkg, duration, id,
                   level=c("dep", "med", "obs"), 
                   cascade=TRUE, 
                   start_end=c("both", "start", "end")){
  level <- match.arg(level)
  start_end <- match.arg(start_end)
  duration <- c(duration, 0)
  n <- length(duration)
  
  if(level=="dep"){
    if(start_end %in% c("both", "start"))
      pkg$dep <- mutate(pkg$dep, start = start + duration[match(depID, id, n)])
    if(start_end %in% c("both", "end"))
      pkg$dep <- mutate(pkg$dep, end = end + duration[match(depID, id, n)])
    if(cascade){
      pkg$med <- mutate(pkg$med, 
                        timestamp = timestamp + duration[match(depID, id, n)])
      pkg$obs <- mutate(pkg$obs, 
                        timestamp = timestamp + duration[match(depID, id, n)])
    }
  } else
    if(level=="med"){
      pkg$med <- mutate(pkg$med, 
                        timestamp = timestamp + duration[match(medID, id, n)])
      if(cascade){
        pkg$obs <- mutate(pkg$obs, 
                          timestamp = timestamp + duration[match(medID, id, n)])
      }
    } else{
      pkg$obs <- mutate(pkg$obs, 
                        timestamp = timestamp + duration[match(obsID, id, n)])
    }
  pkg
}
```

**Comment by damianooldoni** [comment]
Thanks @MarcusRowcliffe for the insights. @sannegovaert's PR (#108) is implementing the function as described by @peterdesmet, so it changes both `deploymentStart` and `deploymentEnd` of a deployment and apply the changes also to observations and media.

In my opinion we should aim to have a set of (narrowly) scoped functions instead of one (too much?) flexible function as such a function will:
- Be complex to debug/test: many encapsulated if-else to deal with combination of args values
- need a lot of checks (assertions) to avoid users create non-sense results which will propagate in all their own analysis afterwards. Examples: modify end or start only with cascade could set obs or media out of the deployment start-end interval. Modify media timestamps (level `med`) but not observations (cascade `FALSE`) could also generate unexpected results at alater stage.

Notice also how camtrapdp is already applying this kind of software dev paradigm: we don't have one `filter()` function, but three: [`filter_deployments()`](https://inbo.github.io/camtrapdp/reference/filter_deployments.html), [`filter_media()`](https://inbo.github.io/camtrapdp/reference/filter_media.html) and [`filter_observations()`](https://inbo.github.io/camtrapdp/reference/filter_observations.html) and each of them has a specific "cascade" behavior which makes sense for users. If users want to filter in other (more dangerous?) ways, they can using dplyr and assignment functions [`deployments(x)<-`](https://inbo.github.io/camtrapdp/reference/deployments.html), [`observations(x)<-`](https://inbo.github.io/camtrapdp/reference/observations.html) or [`media(x)<-`](https://inbo.github.io/camtrapdp/reference/media.html).

I still don't have a perfect solution, but I think that the function `correct_time()` implemented in PR #108 is ok as it is. Maybe we could rename it as `correct_time_deployments()` when a `correct_time_observations()` and `correct_time_media()` will ever be developed. Still, are there situations/use cases where these functions are needed?

In use case #107, only `deploymentEnd` needs to be changed, but there is no need to change timestamps of some obs and media which actually are invalid (=to filter out). @MarcusRowcliffe: are there use cases where only start or end of a deployment need to be changed and timestamps of media/obs need to be changed as well instead of being removed? Are there other interesting use cases we need to take into account?

**Comment by sannegovaert** [comment]
@MarcusRowcliffe This function had now been [implemented](https://github.com/inbo/camtrapdp/pull/108). At the end, it was renamed to `shift_time()` so it is clear the user need to decide how to correct the datetimes.
There is an example on the [pkgdwn website](https://inbo.github.io/camtrapdp/reference/shift_time.html#ref-examples). 

I think part of what you explained of desired functionality (which is not all implemented) is also covered in issue #107.
